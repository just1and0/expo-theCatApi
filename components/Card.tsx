import { useState } from 'react';
import { View, Text, TouchableOpacity, StyleSheet } from 'react-native';
import { Image } from 'expo-image';
import en from '@/constants/en';
import { CardProps } from '@/types';

export function Card({
  cat, onFavourite, onVote
}: CardProps) {

  const [isFavourite, setIsFavourite] = useState(cat.isFavourite);
  const [score, setScore] = useState(cat.score || 0);

  const handleFavourite = () => {
    setIsFavourite(!isFavourite);
    onFavourite(cat.id, !isFavourite);
  };

  const handleVote = (type: 'up' | 'down') => {
    const newScore = type === 'up' ? score + 1 : score - 1;
    setScore(newScore);
    onVote(cat.id, type);
  };

  const blurhash =
    '|rF?hV%2WCj[ayj[a|j[az_NaeWBj@ayfRayfQfQM{M|azj[azf6fQfQfQIpWXofj[ayj[j[fQayWCoeoeaya}j[ayfQa{oLj?j[WVj[ayayj[fQoff7azayj[ayj[j[ayofayayayj[fQj[ayayj[ayfjj[j[ayjuayj[';

  return (
    <View style={styles.card}>
      <Image source={cat.url} style={styles.image} placeholder={{ blurhash }} transition={1000} />
      <View style={styles.info}>
        <Text style={styles.scoreText}>{score}</Text>
      </View>
      <View style={styles.actions}>
        <TouchableOpacity onPress={handleFavourite} style={styles.button}>
          <Text style={styles.buttonText}>{isFavourite ? en.CARD.UNFAVOURITE : en.CARD.FAVOURITE}</Text>
        </TouchableOpacity>
        <View style={styles.voteContainer}>
          <TouchableOpacity onPress={() => handleVote('up')} style={styles.voteButton}>
            <Text style={styles.voteText}>{en.CARD.VOTE_UP}</Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={() => handleVote('down')} style={styles.voteButton}>
            <Text style={styles.voteText}>{en.CARD.VOTE_DOWN}</Text>
          </TouchableOpacity> 
        </View>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  card: {
    width: '90%',
    maxWidth: 400,
    borderWidth: 1,
    borderColor: '#ccc',
    borderRadius: 10,
    padding: 16,
    marginBottom: 20,
    alignSelf: 'center',
    backgroundColor: '#fff',
  },
  image: {
    width: '100%',
    height: 200,
    borderRadius: 8,
    marginBottom: 12,
  },
  info: {
    marginBottom: 12,
    alignItems: 'center',
  },
  scoreText: {
    fontSize: 16,
    fontWeight: 'bold',
    color: '#333',
  },
  actions: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  button: {
    paddingVertical: 8,
    paddingHorizontal: 16,
    borderWidth: 1,
    borderRadius: 5,
    backgroundColor: '#ff7f50',
    borderColor: '#ff7f50',
    marginRight: 12,
  },
  buttonText: {
    color: '#fff',
    fontWeight: 'bold',
  },
  voteContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  voteButton: {
    padding: 8,
    marginHorizontal: 6,
    borderWidth: 1,
    borderColor: '#ccc',
    borderRadius: 5,
  },
  voteText: {
    fontSize: 18,
  },
});
