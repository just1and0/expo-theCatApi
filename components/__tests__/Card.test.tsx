import * as React from 'react';
import renderer from 'react-test-renderer';

import { Card } from '../Card';

it(`renders correctly`, () => {
  const tree = renderer.create(
    <Card cat={{
      id: '1',
      url: 'link-to-image',
      isFavourite: false,
      score: 0
    }} onFavourite={function (id: string, isFavourite: boolean): void {

    }} onVote={function (id: string, type: 'up' | 'down'): void {

    }}
    />).toJSON();

  expect(tree).toMatchSnapshot();
});
