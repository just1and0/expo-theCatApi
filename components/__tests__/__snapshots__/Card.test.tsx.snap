// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    {
      "alignSelf": "center",
      "backgroundColor": "#fff",
      "borderColor": "#ccc",
      "borderRadius": 10,
      "borderWidth": 1,
      "marginBottom": 20,
      "maxWidth": 400,
      "padding": 16,
      "width": "90%",
    }
  }
>
  <ViewManagerAdapter_ExpoImage
    borderRadius={8}
    containerViewRef={"[React.ref]"}
    contentFit="cover"
    contentPosition={
      {
        "left": "50%",
        "top": "50%",
      }
    }
    height={200}
    marginBottom={12}
    nativeViewRef={"[React.ref]"}
    onError={[Function]}
    onLoad={[Function]}
    onLoadStart={[Function]}
    onProgress={[Function]}
    placeholder={
      [
        {
          "height": 16,
          "uri": "blurhash:/%7CrF%3FhV%252WCj%5Bayj%5Ba%7Cj%5Baz_NaeWBj@ayfRayfQfQM%7BM%7Cazj%5Bazf6fQfQfQIpWXofj%5Bayj%5Bj%5BfQayWCoeoeaya%7Dj%5BayfQa%7BoLj%3Fj%5BWVj%5Bayayj%5BfQoff7azayj%5Bayj%5Bj%5Bayofayayayj%5BfQj%5Bayayj%5Bayfjj%5Bj%5Bayjuayj%5B",
          "width": 16,
        },
      ]
    }
    source={
      [
        {
          "uri": "link-to-image",
        },
      ]
    }
    style={
      {
        "borderRadius": 8,
        "height": 200,
        "marginBottom": 12,
        "width": "100%",
      }
    }
    transition={
      {
        "duration": 1000,
      }
    }
    width="100%"
  />
  <View
    style={
      {
        "alignItems": "center",
        "marginBottom": 12,
      }
    }
  >
    <Text
      style={
        {
          "color": "#333",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      0
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#ff7f50",
          "borderColor": "#ff7f50",
          "borderRadius": 5,
          "borderWidth": 1,
          "marginRight": 12,
          "opacity": 1,
          "paddingHorizontal": 16,
          "paddingVertical": 8,
        }
      }
    >
      <Text
        style={
          {
            "color": "#fff",
            "fontWeight": "bold",
          }
        }
      >
        Favourite
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "borderColor": "#ccc",
            "borderRadius": 5,
            "borderWidth": 1,
            "marginHorizontal": 6,
            "opacity": 1,
            "padding": 8,
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 18,
            }
          }
        >
          ⬆️
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "borderColor": "#ccc",
            "borderRadius": 5,
            "borderWidth": 1,
            "marginHorizontal": 6,
            "opacity": 1,
            "padding": 8,
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 18,
            }
          }
        >
          ⬇️
        </Text>
      </View>
    </View>
  </View>
</View>
`;
